include/swift/AST/DiagnosticsSema.def

```diff
@@ -4007,7 +4007,7 @@ WARNING(use_of_qq_on_non_optional_value,none,
         "left side of nil coalescing operator '?""?' has non-optional type %0, "
         "so the right side is never used", (Type))
 WARNING(nonoptional_compare_to_nil,none,
-        "comparing non-optional value of type %0 to 'nil' always returns"
+        "comparing non-optional value of type %0 to 'nil' or 'Optional.none' always returns"
```

Comment:
You can keep a single warning, but you can switch between `nil` or `Optional.none` in the text by adding another bool to the diagnostic. For example:

```
WARNING(nonoptional_compare_to_nil,none,
         "comparing non-optional value of type %0 to 'nil' always returns"
         "comparing non-optional value of type %0 to ‘%select{Optional.none|nil}1’ always returns"
         " %select{false|true}2", (Type, bool, bool))
```

Then you can do:

```c++
Ctx.Diags.diagnose(DRE->getLoc(), diag:: nonoptional_compare_to_nil,
                              subExpr->getType(), isNilLiteral, isTrue)
```

where `isOptionalNil` is `true` when the user has written `nil`, otherwise it’s `falsely (i.e. if they have written `.none`)

Reply:
Thanks for the tips -- updated 👍🏻